### PR TITLE
HIVE-24168: Disable hdfsEncryptionShims cache during query-based comp…

### PR DESCRIPTION
**Pull request by klcopp**
Hive keeps a cache of encryption shims in SessionState (Map<URI, HadoopShims.HdfsEncryptionShim> hdfsEncryptionShims). Each encryption shim in the cache stores a FileSystem object.
After compaction where the session user is not the same user as the owner of the partition/table directory, we close all FileSystem objects associated with the user running the compaction, possibly closing an FS stored in the encryption shim cache. The next time query-based compaction is run on a table/partition owned by the same user, compaction will fail in MoveTask[1] since the FileSystem stored in the cache was closed.
This change disables the cache during query-based compaction (optionally; default: disabled).

[1] Error:
{code:java}
2020-09-08 11:23:50,170 ERROR org.apache.hadoop.hive.ql.Driver: [rncdpdev-2.fyre.ibm.com-27]: FAILED: Execution Error, return code 1 from org.apache.hadoop.hive.ql.exec.MoveTask. java.io.IOException: Filesystem closed. org.apache.hadoop.hive.ql.metadata.HiveException: java.io.IOException: Filesystem closed

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java [hunk 1]
```diff
@@ -445,14 +446,23 @@ public static boolean isCompactionTable(Properties tblProperties) {
...
   public static boolean isCompactionTable(Map<String, String> parameters) {
     return Boolean.valueOf(parameters.getOrDefault(COMPACTOR_TABLE_PROPERTY, "false"));
   }
 
+  /**
+   * Determine if a table is used during query based compaction for MM insert-only tables.
+   * @param parameters table properties map
+   * @return true, if the parameters contains {@link AcidUtils#MM_COMPACTOR_TABLE_PROPERTY}
+   */
+  public static boolean isMmCompactionTable(Map<String, String> parameters) {
```
Comment (pvargacl) [inline ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:462]: Shouldn't isCompactionTable return true in both cases? Isn't it a problem the other places we use this util, that the mmCompactionTable are missed?

Comment (klcopp) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:462]: isCompactionTable logically would be true for both full acid and mm tables, but until now we've only used it to mark tables used for compacting full acid tables. AFAIK we don't want to apply the operations we do on full acid compaction tables to mm compaction tables.
I could rename isCompactionTable() to isFullAcidCompactionTable() for easier reading, would that do?

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java [hunk 2]
```diff
@@ -4475,7 +4481,8 @@ public static boolean moveFile(final HiveConf conf, Path srcf, final Path destf,
         destFs.copyFromLocalFile(srcf, destf);
         return true;
       } else {
-        if (needToCopy(conf, srcf, destf, srcFs, destFs, configuredOwner, isManaged)) {
+        if (needToCopy(conf, srcf, destf, srcFs, destFs, configuredOwner, isManaged, isCompactionTable,
+            isMmCompactionTable)) {
           //copy if across file system or encryption zones.
           LOG.debug("Copying source " + srcf + " to " + destf + " because HDFS encryption zones are different.");
```
Comment (zchovan) [inline ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java:4487]: Would it make sense to include the hashes for the fs instances in the debug logs?

Comment (klcopp) [reply ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java:4487]: I think only in hdfs-common (FS.close() et al) – I opened HADOOP-17267 to track it.

File: ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java (hunk 1 above)
Comment (klcopp) [reply ql/src/java/org/apache/hadoop/hive/ql/io/AcidUtils.java:462]: Done.

**Review by szlta — COMMENTED** [review]
I think there is a fundamental problem here: the encryption shim is caching FS objects, while hadoop-common's FS cache is doing the same as well. The two don't work well together as we've seen.

As of now we have only experienced this with query based compaction, but I'd be surprised if we didn't see this in the future too but for some other use cases: the whole setup is basically a trap, and no developer will know not to call FS.close, as nobody knew either when compaction was added.

I understand that HIVE-24168 only intends to fix this for compaction, but I think given the thoughts above we could have a more general approach:
I see bunch of method signatures changed here for the sake of injecting information about whether what we do is compaction or not, eventually being consumed by needToCopy() only.
I think this is too specific, we could perhaps come up with a new config entry which is runtime only and use that to control whether encryption shim caching should be turned on or not. You can use the same way to calculate this for signalling the presence of compaction, and since needToCopy is given a conf instance anyway, we could use it there only, without changing all those signatures for passing these booleans.
Although the trap is still there, this config entry could be reused in the future for other cases where we want to disable encryption shim caching.

**Comment by klcopp** [comment]
@szlta What you say makes sense. The only issue being that if we have one session-level config controlling the cache, and we use it to always turn off the cache during compaction, we'll never allow turning it back on during compaction – this fix wouldn't be behind a feature flag. I'm tentatively okay with this – @zchovan are you ok with this?

**Review by szlta — APPROVED** [review]

**Pull request merged**